src/main/java/apoc/custom/CypherProcedures.java

```diff
@@ -364,30 +370,76 @@ public static GraphPropertiesProxy getProperties(GraphDatabaseAPI api) {
         }
 
         private void restoreProcedures() {
+            timer = new Timer(this.getClass().getSimpleName());
             CustomStatementRegistry registry = new CustomStatementRegistry(api, log);
-            Map<String, Map<String,Map<String, Object>>> stored = readData(properties);
-            stored.get(FUNCTIONS).forEach((name, data) -> {
-                registry.registerFunction(name, (String) data.get("statement"), (String) data.get("output"),
-                        (List<List<String>>) data.get("inputs"), (Boolean)data.get("forceSingle"), (String) data.get("description"));
-            });
-            stored.get(PROCEDURES).forEach((name, data) -> {
-                registry.registerProcedure(name, (String) data.get("statement"), (String) data.get("mode"),
-                        (List<List<String>>) data.get("outputs"), (List<List<String>>) data.get("inputs"), (String) data.get("description"));
-            });
+            timer.scheduleAtFixedRate(new TimerTask() {
+                private AtomicLong highWaterMark = new AtomicLong(0);
+                private static final String LAST_UPDATE_KEY = "lastUpdate";
+                @Override
+                public void run() {
+                    Map<String, Map<String, Map<String, Object>>> storage = readData(properties);
+                    int updated = storage.entrySet().stream()
+                            .filter(entry -> CypherProcedures.PROCEDURES.equals(entry.getKey()) || CypherProcedures.FUNCTIONS.equals(entry.getKey()))
+                            .flatMap(storageEntry -> {
+                                String type = storageEntry.getKey();
+                                Map<String, Map<String, Object>> map = storageEntry.getValue();
+                                boolean isProcedure = CypherProcedures.PROCEDURES.equals(type);
+                                return map.entrySet().stream()
+                                        // `!e.getValue().containsKey("lastUpdate")` is for backwards compatibility for already present procedures/functions
+                                        .filter(e -> !e.getValue().containsKey(LAST_UPDATE_KEY) || (long) e.getValue().get(LAST_UPDATE_KEY) > highWaterMark.get())
```

Comment:
this whole code block looks too complicated for just handling an potentially absent update key.